base/range.jl

```diff
@@ -47,26 +47,29 @@ function _colon(start::T, step, stop::T) where T
 end
 
 """
-    range(start[, stop]; length, stop, step=1)
+    range(start[, stop]; length, stop, step)
+    range(;start, length, stop, step)
 
-Given a starting value, construct a range either by length or from `start` to `stop`,
-optionally with a given step (defaults to 1, a [`UnitRange`](@ref)).
-One of `length` or `stop` is required.  If `length`, `stop`, and `step` are all specified, they must agree.
-
-If `length` and `stop` are provided and `step` is not, the step size will be computed
-automatically such that there are `length` linearly spaced elements in the range.
-
-If `step` and `stop` are provided and `length` is not, the overall range length will be computed
-automatically such that the elements are `step` spaced.
+Construct a `range` from the arguments.
+Mathematically a range is uniquely determined by any three of `start`, `step`, `stop` and `length`.
+Valid invocations of range are:
+* Call `range` with all four arguments. If the arguments are inconsistent, an error will be thrown.
+* Call `range` with any three of `start`, `step`, `stop`, `length`.
+* Call `range` with two of `start`, `stop`, `length`. In this case `step` will be assumed
+to be one and a [`UnitRange`](@ref) will be returned.
```

Comment:
```suggestion
to be one. If all arguments are integers, a [`UnitRange`](@ref) will be returned.
```
Haven't tried with the PR, but currently:
```
julia> range(1, stop=2.0) isa UnitRange
false
```

Reply:
Yes in my view this contradicts the docs and thus is a bug.

Reply:
Oh I see, I misread the method. This is a nontrivial change, does anything else in Base make non-integer `UnitRange`s?

Reply:
In this PR in the end `range_start_stop(1, 2.0) = start:stop` will be called. So `1:2.0 = 1:1.0:2.0` will be returned.
I was confused and thought `1:2.0` would give a unit range. So your suggestion aligns docs with actual behaviour. 
I guess you are right that non Interger unit ranges are an exotic thing, which we should probably not introduce here.

Reply:
I haven't reviewed all the methods, but this one certainly gives you non-integer UnitRanges:
```
julia> range_start_length(a::Real,          len::Integer) = UnitRange{typeof(a)}(a, oftype(a, a+len-1))
range_start_length (generic function with 1 method)

julia> range_start_length(1.5, 10) |> typeof
UnitRange{Float64}

julia> range_start_length(pi, 10) |> typeof
ERROR: MethodError: no method matching Irrational{:π}(::Float64)
```

Reply:
Mhh I get
```julia
julia> range(1.5, length=10) |> typeof
StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}}

julia> Base.range_start_length(1.5, 10) |> typeof
StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}}

julia> @which Base.range_start_length(1.5, 10)
range_start_length(a::AbstractFloat, len::Integer) in Base at range.jl:137
```
And `pi` indeed errors.

Reply:
`pi` also errors on master.

Reply:
Indeed. that's a bug. But catching the `start=1.5` with a *more* specific method `range_start_length(a::AbstractFloat, len::Integer) = range_start_step_length(a, oftype(a, 1), len)`, and UnitRange as the fallback, strikes me as a bad idea. For example: 
```
julia> range_start_length(1//3, 10) |> typeof
UnitRange{Rational{Int64}}
```
Although surprisingly master does do that too, I'd be tempted to still call it a bug:
```
julia> range(1//3, length=10) |> typeof
UnitRange{Rational{Int64}}

julia> range(1//3, length=10, step=1) |> typeof
StepRangeLen{Rational{Int64}, Rational{Int64}, Int64}
```

Reply:
Yeah, I had similar thoughts, when I saw this. These tricky `range_start_length` methods are copied from master in this PR. If somebody knows what is intentional and what is by accident here, that would be great. Otherwise, I would prefer to just mimic behaviour of master here and address UnitRange and pi etc in a separate issue/PR.

Reply:
I meant to say, I agree that leaving these behaviours alone for now sounds like a sensible choice.